IllegalArgumentException with fail modifier

Reported on Gitter https://gitter.im/olafurpg/mdoc?at=5bb5cdb5435c2a518e4cc235

````
$ cat docs/in.md
```scala mdoc:fail
fs2.Stream.eval(println("Do not ever do this"))
```
$ coursier launch com.geirsson:mdoc_2.12.7:0.5.1 co.fs2:fs2-core_2.12:0.10.4
info: Compiling /Users/ollie/dev/metals/target/docs/in.md
mdoc.internal.cli.FileException: /Users/ollie/dev/metals/target/docs/in.md
Caused by: java.lang.IllegalArgumentException: 6 is not a valid line number, allowed [0..1]
	at scala.meta.internal.inputs.InternalInput.lineToOffset(InternalInput.scala:30)
	at scala.meta.internal.inputs.InternalInput.lineToOffset$(InternalInput.scala:26)
	at scala.meta.inputs.Input$Slice.lineToOffset(Input.scala:83)
	at scala.meta.inputs.Position$Range$.apply(Position.scala:58)
	at mdoc.internal.pos.PositionSyntax$XtensionRangePositionMdoc.toMeta(PositionSyntax.scala:50)
	at mdoc.internal.pos.PositionSyntax$XtensionRangePositionMdoc.formatMessage(PositionSyntax.scala:34)
	at mdoc.internal.markdown.Renderer$.$anonfun$renderEvaluatedSection$2(Renderer.scala:134)
	at scala.collection.immutable.List.foreach(List.scala:388)
	at mdoc.internal.markdown.Renderer$.$anonfun$renderEvaluatedSection$1(Renderer.scala:119)
	at mdoc.internal.markdown.Renderer$.$anonfun$renderEvaluatedSection$1$adapted(Renderer.scala:100)
	at scala.collection.immutable.List.foreach(List.scala:388)
	at mdoc.internal.markdown.Renderer$.renderEvaluatedSection(Renderer.scala:100)
	at mdoc.internal.markdown.MdocPostProcessor.$anonfun$processScalaInputs$2(MdocPostProcessor.scala:82)
	at mdoc.internal.markdown.MdocPostProcessor.$anonfun$processScalaInputs$2$adapted(MdocPostProcessor.scala:72)
	at scala.collection.immutable.List.foreach(List.scala:388)
	at mdoc.internal.markdown.MdocPostProcessor.processScalaInputs(MdocPostProcessor.scala:72)
	at mdoc.internal.markdown.MdocPostProcessor.processDocument(MdocPostProcessor.scala:33)
	at com.vladsch.flexmark.internal.PostProcessorManager.postProcess(PostProcessorManager.java:71)
	at com.vladsch.flexmark.internal.PostProcessorManager.processDocument(PostProcessorManager.java:54)
	at com.vladsch.flexmark.parser.Parser.postProcess(Parser.java:389)
```

Expected output
```scala
$ cat eval.scala
object a {
  fs2.Stream.eval(println("Do not ever do this"))
}
$ scalac -classpath $(coursier fetch co.fs2:fs2-core_2.12:0.10.4 -p) eval.scala
eval.scala:2: error: no type parameters for method eval: (fo: F[O])fs2.Stream[F,O] exist so that it can be applied to arguments (Unit)
 --- because ---
argument expression's type is not compatible with formal parameter type;
 found   : Unit
 required: ?F[?O]
  fs2.Stream.eval(println("Do not ever do this"))
             ^
eval.scala:2: error: type mismatch;
 found   : Unit
 required: F[O]
  fs2.Stream.eval(println("Do not ever do this"))
                         ^
two errors found
```

--- a/build.sbt
+++ b/build.sbt
@@ -30,7 +30,6 @@ name := "mdocRoot"
 skip in publish := true
 val V = new {
   val scalameta = "4.1.0"
-  val scalafix = "0.9.1"
 }
 
 lazy val runtime = project
@@ -70,7 +69,6 @@ lazy val mdoc = project
       "com.lihaoyi" %% "fansi" % "0.2.5",
       "io.methvin" % "directory-watcher" % "0.8.0",
       "me.xdrop" % "fuzzywuzzy" % "1.1.10", // for link hygiene "did you mean?"
-      "ch.epfl.scala" %% "scalafix-core" % V.scalafix,
       // live reload
       "io.undertow" % "undertow-core" % "2.0.13.Final",
       "org.jboss.xnio" % "xnio-nio" % "3.6.5.Final",

--- a/mdoc/src/main/scala/mdoc/internal/cli/MainOps.scala
+++ b/mdoc/src/main/scala/mdoc/internal/cli/MainOps.scala
@@ -15,13 +15,13 @@ import mdoc.internal.livereload.UndertowLiveReload
 import mdoc.internal.markdown.DocumentLinks
 import mdoc.internal.markdown.LinkHygiene
 import mdoc.internal.markdown.Markdown
+import mdoc.internal.pos.DiffUtils
 import metaconfig.Configured
 import scala.meta.Input
 import scala.meta.internal.io.FileIO
 import scala.meta.internal.io.PathIO
 import scala.meta.io.AbsolutePath
 import scala.util.control.NonFatal
-import scalafix.internal.diff.DiffUtils
 
 final class MainOps(
     settings: Settings,

--- a/mdoc/src/main/scala/mdoc/internal/io/ConsoleReporter.scala
+++ b/mdoc/src/main/scala/mdoc/internal/io/ConsoleReporter.scala
@@ -1,17 +1,25 @@
 package mdoc.internal.io
 
+import fansi.Attrs
 import fansi.Color._
 import java.io.PrintStream
 import scala.meta.Position
-import scalafix.internal.util.PositionSyntax._
 import mdoc.Reporter
 import mdoc.internal.pos.PositionSyntax._
 
-class ConsoleReporter(ps: PrintStream) extends Reporter {
+class ConsoleReporter(
+    ps: PrintStream,
+    blue: Attrs = Blue,
+    yellow: Attrs = Yellow,
+    red: Attrs = Red
+) extends Reporter {
 
-  private val myInfo = Blue("info")
-  private val myWarning = Yellow("warning")
-  private val myError = Red("error")
+  def formatMessage(pos: Position, severity: String, message: String): String =
+    pos.formatMessage(severity, message)
+
+  private val myInfo = blue("info")
+  private val myWarning = yellow("warning")
+  private val myError = red("error")
 
   private var myWarnings = 0
   private var myErrors = 0
@@ -34,20 +42,20 @@ class ConsoleReporter(ps: PrintStream) extends Reporter {
     throwable.printStackTrace(ps)
   }
   def error(pos: Position, msg: String): Unit = {
-    error(pos.toUnslicedPosition.formatMessage("error", msg))
+    error(formatMessage(pos.toUnslicedPosition, "error", msg))
   }
   def error(msg: String): Unit = {
     myErrors += 1
     ps.println(myError ++ s": $msg")
   }
   def info(pos: Position, msg: String): Unit = {
-    info(pos.toUnslicedPosition.formatMessage("info", msg))
+    info(formatMessage(pos.toUnslicedPosition, "info", msg))
   }
   def info(msg: String): Unit = {
     ps.println(myInfo ++ s": $msg")
   }
   def warning(pos: Position, msg: String): Unit = {
-    warning(pos.toUnslicedPosition.formatMessage("warning", msg))
+    warning(formatMessage(pos.toUnslicedPosition, "warning", msg))
   }
   def warning(msg: String): Unit = {
     myWarnings += 1

--- a/mdoc/src/main/scala/mdoc/internal/markdown/FailInstrumenter.scala
+++ b/mdoc/src/main/scala/mdoc/internal/markdown/FailInstrumenter.scala
@@ -1,0 +1,33 @@
+package mdoc.internal.markdown
+
+import java.io.ByteArrayOutputStream
+import java.io.PrintStream
+
+final class FailInstrumenter(sections: List[SectionInput], i: Int) {
+  private val out = new ByteArrayOutputStream()
+  private val sb = new PrintStream(out)
+  private val gensym = new Gensym()
+  def instrument(): String = {
+    printAsScript()
+    out.toString
+  }
+  private def printAsScript(): Unit = {
+    sb.println("package repl")
+    sb.println("class Session {")
+    sb.println("  def app() = {")
+    sections.zipWithIndex.foreach {
+      case (section, j) =>
+        if (j > i) ()
+        else {
+          if (section.mod == Modifier.Reset) {
+            val nextApp = gensym.fresh("app", "()")
+            sb.print(s"this.$nextApp\n}\ndef $nextApp: Unit = {\n")
+          }
+          if (j == i || section.mod != Modifier.Fail) {
+            sb.println(section.input.text)
+          }
+        }
+    }
+    sb.println("\n  }\n}")
+  }
+}

--- a/mdoc/src/main/scala/mdoc/internal/markdown/Gensym.scala
+++ b/mdoc/src/main/scala/mdoc/internal/markdown/Gensym.scala
@@ -1,0 +1,10 @@
+package mdoc.internal.markdown
+
+class Gensym() {
+  private var counter = 0
+  def fresh(prefix: String, suffix: String = ""): String = {
+    val name = s"$prefix$counter$suffix"
+    counter += 1
+    name
+  }
+}

--- a/mdoc/src/main/scala/mdoc/internal/markdown/Instrumenter.scala
+++ b/mdoc/src/main/scala/mdoc/internal/markdown/Instrumenter.scala
@@ -14,27 +14,37 @@ class Instrumenter(sections: List[SectionInput]) {
   }
   private val out = new ByteArrayOutputStream()
   private val sb = new PrintStream(out)
-  private var counter = 0
-  private var resets = 0
-  private def freshBinder(): String = {
-    val name = s"res$counter"
-    counter += 1
-    name
-  }
+  val gensym = new Gensym()
   private def printAsScript(): Unit = {
-    sections.foreach { section =>
-      if (section.mod == Modifier.Reset) {
-        resets += 1
-        val nextApp = s"app$resets()"
-        sb.print(s"this.$nextApp\n}\ndef $nextApp: Unit = {\n")
-      }
-      sb.println("\n$doc.startSection();")
-      section.source.stats.foreach { stat =>
-        sb.println(s"$$doc.startStatement(${position(stat.pos)});")
-        printStatement(stat, section.mod, sb)
-        sb.println("\n$doc.endStatement();")
-      }
-      sb.println("$doc.endSection();")
+    sections.zipWithIndex.foreach {
+      case (section, i) =>
+        if (section.mod == Modifier.Reset) {
+          val nextApp = gensym.fresh("app", "()")
+          sb.print(s"this.$nextApp\n}\ndef $nextApp: Unit = {\n")
+        }
+        sb.println("\n$doc.startSection();")
+        if (section.mod == Modifier.Fail) {
+          sb.println(s"$$doc.startStatement(${position(section.source.pos)});")
+          val out = new FailInstrumenter(sections, i).instrument()
+          val literal = Instrumenter.stringLiteral(out)
+          val binder = gensym.fresh("res")
+          sb.append("val ")
+            .append(binder)
+            .append(" = _root_.mdoc.internal.document.Macros.Delay(")
+            .append(literal)
+            .append(", ")
+            .append(position(section.source.pos))
+            .append(");")
+          printBinder(binder, section.source.pos)
+          sb.println("\n$doc.endStatement();")
+        } else {
+          section.source.stats.foreach { stat =>
+            sb.println(s"$$doc.startStatement(${position(stat.pos)});")
+            printStatement(stat, section.mod, sb)
+            sb.println("\n$doc.endStatement();")
+          }
+        }
+        sb.println("$doc.endSection();")
     }
   }
 
@@ -48,7 +58,7 @@ class Instrumenter(sections: List[SectionInput]) {
         case Binders(names) =>
           names.map(name => name -> name.pos)
         case _ =>
-          val fresh = freshBinder()
+          val fresh = gensym.fresh("res")
           sb.print(s"val $fresh = ")
           List(Name(fresh) -> stat.pos)
       }
@@ -60,7 +70,7 @@ class Instrumenter(sections: List[SectionInput]) {
 
     case Modifier.Fail =>
       val literal = Instrumenter.stringLiteral(stat.pos.text)
-      val binder = freshBinder()
+      val binder = gensym.fresh("res")
       sb.append("val ")
         .append(binder)
         .append(" = _root_.mdoc.internal.document.Macros.fail(")

--- a/mdoc/src/main/scala/mdoc/internal/markdown/MarkdownCompiler.scala
+++ b/mdoc/src/main/scala/mdoc/internal/markdown/MarkdownCompiler.scala
@@ -1,6 +1,8 @@
 package mdoc.internal.markdown
 
+import java.io.ByteArrayOutputStream
 import java.io.File
+import java.io.PrintStream
 import java.net.URL
 import java.net.URLClassLoader
 import java.nio.file.Path
@@ -19,7 +21,8 @@ import mdoc.Reporter
 import mdoc.document.Document
 import mdoc.document._
 import mdoc.internal.document.DocumentBuilder
-import scalafix.internal.util.PositionSyntax._
+import mdoc.internal.pos.PositionSyntax
+import scala.reflect.internal.util.{Position => GPosition}
 import mdoc.internal.pos.TokenEditDistance
 import mdoc.internal.pos.PositionSyntax._
 
@@ -120,53 +123,79 @@ class MarkdownCompiler(
     case _ =>
   }
 
+  private def toSource(input: Input): BatchSourceFile =
+    new BatchSourceFile(input.syntax, new String(input.chars))
+
+  def fail(original: Seq[Tree], input: Input): String = {
+    sreporter.reset()
+    val run = new global.Run
+    run.compileSources(List(toSource(input)))
+    val out = new ByteArrayOutputStream()
+    val ps = new PrintStream(out)
+    val edit = TokenEditDistance.toTokenEdit(original, input)
+    sreporter.infos.foreach {
+      case sreporter.Info(pos, msgOrNull, gseverity) =>
+        val msg = nullableMessage(msgOrNull)
+        val mpos = toMetaPosition(edit, pos)
+        val severity = gseverity.toString().toLowerCase
+        val formatted = PositionSyntax.formatMessage(mpos, severity, msg, includePath = false)
+        ps.println(formatted)
+    }
+    out.toString()
+  }
+
   def compile(input: Input, vreporter: Reporter, edit: TokenEditDistance): Option[ClassLoader] = {
     clearTarget()
     sreporter.reset()
     val run = new global.Run
-    val label = input match {
-      case Input.File(path, _) => path.toString()
-      case Input.VirtualFile(path, _) => path
-      case _ => "(input)"
-    }
-    run.compileSources(List(new BatchSourceFile(label, new String(input.chars))))
+    run.compileSources(List(toSource(input)))
     if (!sreporter.hasErrors) {
       Some(new AbstractFileClassLoader(target, appClassLoader))
     } else {
-      def toOffsetPosition(offset: Int): Position = {
-        edit.toOriginal(offset) match {
-          case Left(_) =>
-            Position.None
-          case Right(p) =>
-            p.toUnslicedPosition
-        }
-      }
-      sreporter.infos.foreach {
-        case sreporter.Info(pos, msgOrNull, severity) =>
-          val msg = if (msgOrNull == null) "" else msgOrNull
-          val mpos =
-            if (pos.isDefined) {
-              if (pos.isRange) {
-                (edit.toOriginal(pos.start), edit.toOriginal(pos.end - 1)) match {
-                  case (Right(start), Right(end)) =>
-                    Position.Range(start.input, start.start, end.end).toUnslicedPosition
-                  case (_, _) =>
-                    toOffsetPosition(pos.point)
-                }
-              } else {
-                toOffsetPosition(pos.point)
-              }
-            } else {
-              Position.None
-            }
-          severity match {
-            case sreporter.ERROR => vreporter.error(mpos, msg)
-            case sreporter.INFO => vreporter.info(mpos, msg)
-            case sreporter.WARNING => vreporter.warning(mpos, msg)
-          }
-        case _ =>
-      }
+      report(vreporter, edit)
       None
     }
   }
+
+  private def toMetaPosition(edit: TokenEditDistance, pos: GPosition): Position = {
+    def toOffsetPosition(offset: Int): Position = {
+      edit.toOriginal(offset) match {
+        case Left(_) =>
+          Position.None
+        case Right(p) =>
+          p.toUnslicedPosition
+      }
+    }
+    if (pos.isDefined) {
+      if (pos.isRange) {
+        (edit.toOriginal(pos.start), edit.toOriginal(pos.end - 1)) match {
+          case (Right(start), Right(end)) =>
+            Position.Range(start.input, start.start, end.end).toUnslicedPosition
+          case (_, _) =>
+            toOffsetPosition(pos.point)
+        }
+      } else {
+        toOffsetPosition(pos.point)
+      }
+    } else {
+      Position.None
+    }
+  }
+
+  private def nullableMessage(msgOrNull: String): String =
+    if (msgOrNull == null) "" else msgOrNull
+  private def report(vreporter: Reporter, edit: TokenEditDistance): Unit = {
+    sreporter.infos.foreach {
+      case sreporter.Info(pos, msgOrNull, severity) =>
+        val msg = nullableMessage(msgOrNull)
+        val mpos = toMetaPosition(edit, pos)
+        severity match {
+          case sreporter.ERROR => vreporter.error(mpos, msg)
+          case sreporter.INFO => vreporter.info(mpos, msg)
+          case sreporter.WARNING => vreporter.warning(mpos, msg)
+        }
+      case _ =>
+    }
+  }
+
 }

--- a/mdoc/src/main/scala/mdoc/internal/markdown/MdocPostProcessor.scala
+++ b/mdoc/src/main/scala/mdoc/internal/markdown/MdocPostProcessor.scala
@@ -86,7 +86,8 @@ class MdocPostProcessor(implicit ctx: Context) extends DocumentPostProcessor {
           rendered,
           section,
           ctx.reporter,
-          ctx.settings.variablePrinter
+          ctx.settings.variablePrinter,
+          ctx.compiler
         )
         mod match {
           case Modifier.Silent =>

--- a/mdoc/src/main/scala/mdoc/internal/markdown/Renderer.scala
+++ b/mdoc/src/main/scala/mdoc/internal/markdown/Renderer.scala
@@ -7,6 +7,8 @@ import mdoc.Variable
 import mdoc.document.CompileResult
 import mdoc.document.CrashResult
 import mdoc.document.CrashResult.Crashed
+import mdoc.document.RangePosition
+import mdoc.internal.document.Macros
 import mdoc.internal.document.MdocExceptions
 import mdoc.internal.pos.PositionSyntax._
 import mdoc.internal.pos.TokenEditDistance
@@ -29,7 +31,7 @@ object Renderer {
       MarkdownCompiler.buildDocument(compiler, reporter, inputs, instrumented, filename)
     doc.sections
       .map(s => s"""```scala
-                   |${Renderer.renderEvaluatedSection(doc, s, reporter, printer)}
+                   |${Renderer.renderEvaluatedSection(doc, s, reporter, printer, compiler)}
                    |```""".stripMargin)
       .mkString("\n")
   }
@@ -90,13 +92,17 @@ object Renderer {
       doc: EvaluatedDocument,
       section: EvaluatedSection,
       reporter: Reporter,
-      printer: Variable => String
+      printer: Variable => String,
+      compiler: MarkdownCompiler
   ): String = {
     val baos = new ByteArrayOutputStream()
     val sb = new PrintStream(baos)
     val stats = section.source.stats.lift
     val input = section.source.pos.input
     val totalStats = section.source.stats.length
+    if (section.mod.isFail) {
+      sb.print(section.input.text)
+    }
     section.section.statements.zip(section.source.stats).zipWithIndex.foreach {
       case ((statement, tree), statementIndex) =>
         val pos = tree.pos
@@ -107,10 +113,14 @@ object Renderer {
             previousStatement.pos.end
         }
         val leadingTrivia = Position.Range(input, leadingStart, pos.start)
-        sb.append(leadingTrivia.text)
+        if (!section.mod.isFail) {
+          sb.append(leadingTrivia.text)
+        }
         val endOfLinePosition =
           Position.Range(pos.input, pos.startLine, pos.startColumn, pos.endLine, Int.MaxValue)
-        sb.append(endOfLinePosition.text)
+        if (!section.mod.isFail) {
+          sb.append(endOfLinePosition.text)
+        }
         if (statement.out.nonEmpty) {
           sb.append("\n")
           appendFreshMultiline(sb, statement.out)
@@ -132,6 +142,17 @@ object Renderer {
                     appendFreshMultiline(sb, tpos.formatMessage(section, msg))
                   case CompileResult.TypeError(msg, tpos) =>
                     appendFreshMultiline(sb, tpos.formatMessage(section, msg))
+                  case Macros.Delay(instrumented, startLine, startColumn, endLine, endColumn) =>
+                    val compiled =
+                      compiler.fail(doc.sections.map(_.source), Input.String(instrumented))
+                    if (compiled.isEmpty) {
+                      val tpos = new RangePosition(startLine, startColumn, endLine, endColumn)
+                      reporter.error(
+                        tpos.toMeta(section),
+                        s"Expected compile error but statement type-checked successfully"
+                      )
+                    }
+                    appendFreshMultiline(sb, compiled)
                   case _ =>
                     val obtained = pprint.PPrinter.BlackWhite.apply(binder).toString()
                     throw new IllegalArgumentException(

--- a/mdoc/src/main/scala/mdoc/internal/pos/DiffUtils.scala
+++ b/mdoc/src/main/scala/mdoc/internal/pos/DiffUtils.scala
@@ -1,0 +1,24 @@
+package mdoc.internal.pos
+
+import difflib.{DiffUtils => DU}
+import scala.collection.JavaConverters._
+
+object DiffUtils {
+  def unifiedDiff(
+      originalFileName: String,
+      revisedFileName: String,
+      originalLines: List[String],
+      revisedLines: List[String],
+      contextSize: Int
+  ): String = {
+    val patch = DU.diff(originalLines.asJava, revisedLines.asJava)
+    val diff = DU.generateUnifiedDiff(
+      originalFileName,
+      revisedFileName,
+      originalLines.asJava,
+      patch,
+      contextSize
+    )
+    diff.asScala.mkString("\n")
+  }
+}

--- a/mdoc/src/main/scala/mdoc/internal/pos/PositionSyntax.scala
+++ b/mdoc/src/main/scala/mdoc/internal/pos/PositionSyntax.scala
@@ -5,7 +5,6 @@ import scala.meta.Input
 import scala.meta.Position
 import scala.meta.io.AbsolutePath
 import scala.meta.io.RelativePath
-import scalafix.internal.util.PositionSyntax._
 import mdoc.document.RangePosition
 import mdoc.internal.cli.Settings
 import mdoc.internal.markdown.EvaluatedSection
@@ -84,4 +83,75 @@ object PositionSyntax {
       }
     }
   }
+
+  def formatMessage(
+      pos: Position,
+      severity: String,
+      message: String,
+      includePath: Boolean = true
+  ): String =
+    pos match {
+      case Position.None =>
+        s"$severity: $message"
+      case _ =>
+        new java.lang.StringBuilder()
+          .append(if (includePath) pos.lineInput else "")
+          .append(if (!includePath || severity.isEmpty) "" else " ")
+          .append(severity)
+          .append(
+            if (message.isEmpty) ""
+            else if (severity.isEmpty) " "
+            else if (message.startsWith("\n")) ":"
+            else ": "
+          )
+          .append(message)
+          .append("\n")
+          .append(pos.lineContent)
+          .append("\n")
+          .append(pos.lineCaret)
+          .toString
+    }
+
+  implicit class XtensionPositionsScalafix(private val pos: Position) extends AnyVal {
+
+    def contains(other: Position): Boolean = {
+      pos.start <= other.start &&
+      pos.end >= other.end
+    }
+
+    def formatMessage(severity: String, message: String): String =
+      PositionSyntax.formatMessage(pos, severity, message)
+
+    /** Returns a formatted string of this position including filename/line/caret. */
+    def lineInput: String =
+      s"${pos.input.syntax}:${pos.startLine + 1}:${pos.startColumn + 1}:"
+
+    def rangeNumber: String =
+      s"${pos.startLine + 1}:${pos.startColumn + 1}..${pos.endLine + 1}:${pos.endColumn + 1}"
+
+    def lineCaret: String = pos match {
+      case Position.None =>
+        ""
+      case _ =>
+        val caret =
+          if (pos.start == pos.end) "^"
+          else if (pos.startLine == pos.endLine) "^" * (pos.end - pos.start)
+          else "^"
+        (" " * pos.startColumn) + caret
+    }
+
+    def lineContent: String = pos match {
+      case Position.None => ""
+      case range: Position.Range =>
+        val pos = Position.Range(
+          range.input,
+          range.startLine,
+          0,
+          range.startLine,
+          Int.MaxValue
+        )
+        pos.text
+    }
+  }
+
 }

--- a/mdoc/src/main/scala/mdoc/internal/pos/TokenEditDistance.scala
+++ b/mdoc/src/main/scala/mdoc/internal/pos/TokenEditDistance.scala
@@ -116,7 +116,7 @@ object TokenEditDistance {
     }
     val deltas = {
       import scala.collection.JavaConverters._
-      DiffUtils
+      difflib.DiffUtils
         .diff(original.asJava, revised.asJava, TokenEqualizer)
         .getDeltas
         .iterator()

--- a/readme.md
+++ b/readme.md
@@ -274,11 +274,11 @@ After:
 ````
 ```scala
 val x: Int = ""
-// type mismatch;
+// error: type mismatch;
 //  found   : String("")
 //  required: Int
 // val x: Int = ""
-//              ^
+//              ^^
 ```
 ````
 
@@ -413,12 +413,12 @@ implicit val x: Int = 41
 implicit val y: Int = 42
 // y: Int = 42
 implicitly[Int] // x is no longer in scope
-// res0: Int = 42
+// res1: Int = 42
 ```
 
 ```scala
 println(x)
-// not found: value x
+// error: not found: value x
 // println(x)
 //         ^
 ```

--- a/runtime/src/main/scala/mdoc/internal/document/Macros.scala
+++ b/runtime/src/main/scala/mdoc/internal/document/Macros.scala
@@ -9,7 +9,13 @@ import mdoc.document.CompileResult._
 import mdoc.document.RangePosition
 
 object Macros {
-
+  case class Delay(
+      code: String,
+      startLine: Int,
+      startColumn: Int,
+      endLine: Int,
+      endColumn: Int
+  )
   def fail(
       code: String,
       startLine: Int,

--- a/tests/unit/src/test/scala/tests/markdown/FailSuite.scala
+++ b/tests/unit/src/test/scala/tests/markdown/FailSuite.scala
@@ -11,15 +11,14 @@ class FailSuite extends BaseMarkdownSuite {
       |val x: Int = "String"
       |```
     """.stripMargin,
-    """
-      |```scala
-      |val x: Int = "String"
-      |// type mismatch;
-      |//  found   : String("String")
-      |//  required: Int
-      |// val x: Int = "String"
-      |//              ^
-      |```
+    """|```scala
+       |val x: Int = "String"
+       |// error: type mismatch;
+       |//  found   : String("String")
+       |//  required: Int
+       |// val x: Int = "String"
+       |//              ^^^^^^^^
+       |```
     """.stripMargin
   )
 
@@ -37,7 +36,7 @@ class FailSuite extends BaseMarkdownSuite {
       |val y: Int = '''Triplequote
       |newlines
       |'''
-      |// type mismatch;
+      |// error: type mismatch;
       |//  found   : String("Triplequote\nnewlines\n")
       |//  required: Int
       |// val y: Int = '''Triplequote
@@ -103,20 +102,19 @@ class FailSuite extends BaseMarkdownSuite {
       |val x: Int = "String"
       |```
     """.stripMargin,
-    """
-      |```scala
-      |println(42)
-      |// 42
-      |```
-      |
-      |```scala
-      |val x: Int = "String"
-      |// type mismatch;
-      |//  found   : String("String")
-      |//  required: Int
-      |// val x: Int = "String"
-      |//              ^
-      |```
+    """|```scala
+       |println(42)
+       |// 42
+       |```
+       |
+       |```scala
+       |val x: Int = "String"
+       |// error: type mismatch;
+       |//  found   : String("String")
+       |//  required: Int
+       |// val x: Int = "String"
+       |//              ^^^^^^^^
+       |```
     """.stripMargin
   )
 
@@ -132,13 +130,76 @@ class FailSuite extends BaseMarkdownSuite {
     // We should reconsider the architecture for the `fail` modifier.
     """|```scala
        |fs2.Stream.eval(println("Do not ever do this"))
-       |// type mismatch;
+       |// error: no type parameters for method eval: (fo: F[O])fs2.Stream[F,O] exist so that it can be applied to arguments (Unit)
+       |//  --- because ---
+       |// argument expression's type is not compatible with formal parameter type;
        |//  found   : Unit
        |//  required: ?F[?O]
-       |// Note that implicit conversions are not applicable because they are ambiguous:
-       |//  both method ArrowAssoc in object Predef of type [A](self: A)ArrowAssoc[A]
-       |//  and method Ensuring in object Predef of type [A](self: A)Ensuring[A]
-       |//  are possible conversion functions from Unit to ?F[?O]
+       |// fs2.Stream.eval(println("Do not ever do this"))
+       |// ^^^^^^^^^^^^^^^
+       |// error: type mismatch;
+       |//  found   : Unit
+       |//  required: F[O]
+       |// fs2.Stream.eval(println("Do not ever do this"))
+       |//                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+       |```
+    """.stripMargin
+  )
+
+  check(
+    "double",
+    """
+      |```scala mdoc:fail
+      |println(a)
+      |println(b)
+      |```
+    """.stripMargin,
+    """|```scala
+       |println(a)
+       |println(b)
+       |// error: not found: value a
+       |// println(a)
+       |//         ^
+       |// error: not found: value b
+       |// println(b)
+       |//         ^
+       |```
+    """.stripMargin
+  )
+
+  check(
+    "edit",
+    """
+      |```scala mdoc:fail
+      |val x = 1
+      |println(a)
+      |println(b)
+      |```
+      |
+      |```scala mdoc:fail
+      |val x = 1
+      |println(c)
+      |println(d)
+      |```
+    """.stripMargin,
+    """|```scala
+       |val x = 1
+       |println(a)
+       |println(b)
+       |// error: not found: value a
+       |// error: not found: value b
+       |```
+       |
+       |```scala
+       |val x = 1
+       |println(c)
+       |println(d)
+       |// error: not found: value c
+       |// println(c)
+       |//         ^
+       |// error: not found: value d
+       |// println(d)
+       |//         ^
        |```
     """.stripMargin
   )

--- a/tests/unit/src/test/scala/tests/markdown/ResetSuite.scala
+++ b/tests/unit/src/test/scala/tests/markdown/ResetSuite.scala
@@ -32,7 +32,7 @@ class ResetSuite extends BaseMarkdownSuite {
        |
        |```scala
        |println(x)
-       |// not found: value x
+       |// error: not found: value x
        |// println(x)
        |//         ^
        |```

